### PR TITLE
Access Wideners

### PR DESCRIPTION
**Pull request by TheGlitch76**
Needs testing but reviews now would be helpful.
This strips the old (unused) gloom system. If someone really really wants gloom-style field ATs, they can go into git history to restore them.

### CODE REVIEW

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 1]
```diff
@@ -153,21 +150,15 @@ private ForgeModJar parseModManifest(Path jarPath) throws IOException, URISyntax
...
-				accessTransformers = AccessTransformerList.parse(fs.getPath("/META-INF/accesstransformer.cfg"));
-			} catch (Exception e) {
-				LOGGER.throwing(Level.ERROR, new RuntimeException("Unable to parse access transformer list", e));
-			}
-
-			if (accessTransformers == null) {
-				accessTransformers = new AccessTransformerList(new ArrayList<>());
+				at = ForgeAccessTransformer.parse(fs.getPath("/META-INF/accesstransformer.cfg"));
+			} catch (ManifestParseException ex) {
+				at = null;
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: The exception shouldn't be silently swallowed here, there should be better error handling / checking (such as whether it exists at all)

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 2]
```diff
@@ -258,13 +259,21 @@ private void transformMod(ForgeModJar forgeModJar) throws IOException, URISyntax
...
 		}
 
 		Files.write(fabricModJson, json.getBytes(StandardCharsets.UTF_8));
 
-		LOGGER.trace("fabric.mod.json: " + json);
+		if (at != null) {
+			Files.write(fs.getPath("/" + accessWidenerName), AccessTransformerConverter.convertToWidener(at, intermediaryHolder));
+		}
+
+		//LOGGER.trace("fabric.mod.json: " + json);
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: Should either uncomment this or remove it

File: src/main/java/com/patchworkmc/PatchworkUI.java [hunk 3]
```diff
@@ -430,22 +430,16 @@ private static void startPatching() throws IOException {
...
 
 		File voldemapBridged = new File(root, "data/mappings/voldemap-bridged-" + version + ".tiny");
-		IMappingProvider bridged;
 
 		if (!voldemapBridged.exists()) {
 			System.out.println("Generating bridged (srg -> intermediary) tiny mappings...");
 
 			TinyWriter tinyWriter = new TinyWriter("srg", "intermediary");
-			bridged = new BridgedMappings(mappings, intermediary);
-			bridged.load(tinyWriter);
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/PatchworkUI.java]: This seems like it would break things...

File: src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	private final IMappingProvider inverted;
+
+	private boolean loaded = false;
+
+	public IntermediaryHolder(IMappingProvider inverted) {
+		this.inverted = inverted;
+	}
+
+	public Map<String, Map<String, Member>> getMappings() {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java]: Since this is only ever used to get the mappings for a given class, we could make this `getMappings(String owner)` instead.

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			} else if (methods != null) {
+				writeMethod(sb, owner, member.name, member.descriptor, methods, false);
+			}
+		}
+	}
+
+	private static void writeField(StringBuilder sb, TransformedField transformed, IntermediaryHolder holder) {
+		String owner = transformed.getOwner();
+		String name = transformed.getName();
+		String descriptor = holder.getMappings().get(owner).get(name).descriptor;
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: Potential NPE for a missing class or field

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		AccessTransformerConverter.writeField(sb, owner, name, descriptor, transformed, true);
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: Are you sure that the qualification is needed?
```suggestion
		writeField(sb, owner, name, descriptor, transformed, true);
```

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+			sb.append('\t').append(owner).append('\t').append("*\n");
+		}
+		if (methods != null) {
+			writeDebugModifier(sb, methods);
+			sb.append('\t').append(owner).append('\t').append("*\n");
+		}
+
+
+		//// Widener
+		for (Map.Entry<String, IntermediaryHolder.Member> entry : holder.getMappings().get(owner).entrySet()) {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: Potential NPE for a missing class

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		sb.append("accessWidener\tv1\t").append("intermediary").append('\n');
+		sb.append("#\tAutomatically generated by Patchwork Patcher.\n");
+		sb.append("#\tThe original (but remapped) AT entry is roughly preserved here for debugging.\n");
+		sb.append("#\tThey may be stripped in future releases.\n\n");
+
+		for (TransformedClass targetClass : accessTransformer.getClasses()) {
+			if (targetClass.getAccessLevel() != AccessLevel.KEEP || targetClass.getFinalization() != Finalization.KEEP) {
+				writeClass(sb, targetClass);
+			}
+			writeWildcards(sb, targetClass.getName(), (TransformedWildcardMember) targetClass.getDefaultForFields(),
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: I don't like the casts here, is there a way to avoid this? Maybe patchwork-manifest should have a better interface here

**Review by coderbot16 — CHANGES_REQUESTED** [review]
This overall looks quite good, but I've noticed a few issues.

**Comment by coderbot16** [comment]
It's removed too

File: src/main/java/com/patchworkmc/patch/StringConstantRemapper.java [hunk 9]
```diff
@@ -7,16 +7,16 @@
...
 /**
  * Remaps Strings into Intermediary. Necessary for things like reflection and ObfuscationRemapperHelper,
  * unless we put SRG and tiny-remapper on the classpath at runtime.
  */
 public class StringConstantRemapper extends ClassVisitor {
-	private NaiveRemapper remapper;
+	private PatchworkRemapper remapper;
 
-	public StringConstantRemapper(ClassVisitor classVisitor, NaiveRemapper remapper) {
+	public StringConstantRemapper(ClassVisitor classVisitor, PatchworkRemapper remapper) {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/patch/StringConstantRemapper.java]: Doesn't it make sense to just take `PatchworkRemapper.Naive` as an argument since that's all that this class uses?

File: src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import net.fabricmc.tinyremapper.IMappingProvider;
+
+/**
+ * A lazily loading representation of all Minecraft names and descriptors.
+ */
+public class IntermediaryHolder {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java]: `IntermediaryHolder` isn't the best name since this isn't necessarily specific to intermediary. Where this is used could be documented as well.

File: src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+		// We load lazily here so that if this class isn't needed we can save some time and memory
+		if (!this.loaded) {
+			inverted.load(new Acceptor());
+			loaded = true;
+		}
+
+		return mappings.get(owner);
+	}
+
+	// kotlin-style object to hold fields.
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java]: a better way to refer to this is `POJO` (plain old java object)
```suggestion
	// kotlin-style object to hold fields and methods.
```

File: src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import net.fabricmc.tinyremapper.IMappingProvider;
+
+/**
+ * A lazily loading representation of all Minecraft names and descriptors.
+ */
+public class IntermediaryHolder {
+	private final HashMap<String, Map<String, Member>> mappings = new HashMap<>();
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/IntermediaryHolder.java]: I think it would make sense to split this between methods and fields, using 2 maps instead of 1.

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 13]
```diff
@@ -65,18 +68,20 @@
...
-	public Patchwork(Path inputDir, Path outputDir, Path dataDir, Path tempDir, IMappingProvider primaryMappings, List<IMappingProvider> devMappings) {
+	public Patchwork(Path inputDir, Path outputDir, Path dataDir, Path tempDir, Path bridgedMappings, List<IMappingProvider> devMappings) {
 		this.inputDir = inputDir;
 		this.outputDir = outputDir;
 		this.dataDir = dataDir;
 		this.tempDir = tempDir;
 		this.clientJarSrg = dataDir.resolve(version + "-client+srg.jar");
-		this.primaryMappings = primaryMappings;
+		this.primaryMappings = TinyUtils.createTinyMappingProvider(bridgedMappings, "srg", "intermediary");
+		this.intermediaryHolder = new IntermediaryHolder(TinyUtils.createTinyMappingProvider(bridgedMappings, "intermediary", "srg"));
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: Seems like this can just as easily be done with the base intermediary mappings instead of having to use bridged mappings?

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 14]
```diff
@@ -65,18 +68,20 @@
...
-	private FieldDescriptorProvider fieldDescriptorProvider;
-	private NaiveRemapper naiveRemapper;
-	private AccessTransformerRemapper accessTransformerRemapper;
+	private PatchworkRemapper patchworkRemapper;
+	private Remapper accessTransformerRemapper;
+	private final IntermediaryHolder intermediaryHolder;
 	private boolean closed = false;
 
-	public Patchwork(Path inputDir, Path outputDir, Path dataDir, Path tempDir, IMappingProvider primaryMappings, List<IMappingProvider> devMappings) {
+	public Patchwork(Path inputDir, Path outputDir, Path dataDir, Path tempDir, Path bridgedMappings, List<IMappingProvider> devMappings) {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: I think we should continue passing the IMappingProvider here. Maybe have another parameter for the intermediary mappings per the comment below.

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 15]
```diff
@@ -153,39 +153,37 @@ private ForgeModJar parseModManifest(Path jarPath) throws IOException, URISyntax
...
-				accessTransformers = AccessTransformerList.parse(fs.getPath("/META-INF/accesstransformer.cfg"));
-			} catch (Exception e) {
-				LOGGER.throwing(Level.ERROR, new RuntimeException("Unable to parse access transformer list", e));
-			}
-
-			if (accessTransformers == null) {
-				accessTransformers = new AccessTransformerList(new ArrayList<>());
+				at = ForgeAccessTransformer.parse(fs.getPath("/META-INF/accesstransformer.cfg"));
+			} catch (ManifestParseException ex) {
+				if (ex.getCause() instanceof NoSuchFileException) {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: Might be better to just check with `Files.exists()` first

File: src/main/java/com/patchworkmc/Patchwork.java [hunk 16]
```diff
@@ -395,25 +406,22 @@ public static void main(String[] args) throws Exception {
...
 			LOGGER.trace("Using cached bridged (srg -> intermediary) tiny mappings");
-
-			bridged = TinyUtils.createTinyMappingProvider(voldemapBridged.toPath(), "srg", "intermediary");
 		}
 
 		Path inputDir = Files.createDirectories(currentPath.resolve("input"));
 		Path outputDir = Files.createDirectories(currentPath.resolve("output"));
 		Path tempDir = Files.createTempDirectory(currentPath, "temp");
-		new Patchwork(inputDir, outputDir, currentPath.resolve("data/"), tempDir, bridged, Collections.emptyList()).patchAndFinish();
+		new Patchwork(inputDir, outputDir, currentPath.resolve("data/"), tempDir, voldemapBridged.toPath(), Collections.emptyList()).patchAndFinish();
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/Patchwork.java]: Why is this change needed?

File: src/main/java/com/patchworkmc/PatchworkUI.java [hunk 17]
```diff
@@ -430,22 +430,17 @@ private static void startPatching() throws IOException {
...
-
+			new BridgedMappings(mappings, intermediary).load(tinyWriter);
 			Files.write(voldemapBridged.toPath(), tinyWriter.toString().getBytes(StandardCharsets.UTF_8));
 
 			System.out.println("Using generated bridged (srg -> intermediary) tiny mappings");
 		} else {
 			System.out.println("Using cached bridged (srg -> intermediary) tiny mappings");
-
-			bridged = TinyUtils.createTinyMappingProvider(voldemapBridged.toPath(), "srg", "intermediary");
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/PatchworkUI.java]: Same concern as with the CLI code.

File: src/main/java/com/patchworkmc/jar/ForgeModJar.java [hunk 18]
```diff
@@ -1,53 +1,23 @@
...
-		this.depedencies = new HashSet<>();
-		// Store all of our dependencies into one Set
-		this.manifest.getDependencyMapping().forEach((modId, list) ->
-					list.forEach((modManifestDependency ->
-							this.depedencies.add(modManifestDependency.getModId()))));
+	public ForgeModJar(Path jarPath, ModManifest manifest) {
+		this(jarPath, manifest, null);
 	}
 
-	public void addDependencyJars(List<ForgeModJar> modJars) {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/jar/ForgeModJar.java:30]: This is okay for now but at some point in the future we'll need to merge ATs from dependency JARs.

File: src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+
+import net.fabricmc.tinyremapper.IMappingProvider;
+
+import net.patchworkmc.manifest.accesstransformer.v2.exception.MissingMappingException;
+
+import com.patchworkmc.Patchwork;
+
+public class PatchworkRemapper {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java:13]: What's the rationale for not using tiny-remapper's Remapper here? (For PatchworkRemapper itself, not PatchworkRemapper.Naive)

File: src/main/java/com/patchworkmc/mapping/remapper/ManifestRemapperImpl.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+	@Override
+	public String remapFieldName(String owner, String name, String descriptor) throws MissingMappingException {
+		owner = owner.replace('.', '/');
+
+		if (!name.startsWith("field_")) {
+			return name;
+		}
+
+		return patchworkRemapper.getField(owner, name);
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/remapper/ManifestRemapperImpl.java]: The fatal exception is never converted back to a missingmapping exception

File: src/main/java/com/patchworkmc/mapping/remapper/ManifestRemapperImpl.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	@Override
+	public String remapMethodName(String owner, String name, String descriptor) throws MissingMappingException {
+		return patchworkRemapper.getMethod(owner, name, descriptor);
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/remapper/ManifestRemapperImpl.java]: Ditto

**Review by coderbot16 — CHANGES_REQUESTED** [review]
This will be a nice addition, but I do currently see some larger issues with this PR currently.

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java (hunk 5 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: Still appears to be an issue

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java (hunk 7 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: Same here

File: src/main/java/com/patchworkmc/jar/ForgeModJar.java (hunk 18 above)
Comment (TheGlitch76) [reply src/main/java/com/patchworkmc/jar/ForgeModJar.java:30]: Why would that be needed? Deps would have their ATs applied to Minecraft, and the mod will be happy.

File: src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+		for (Map.Entry<String, IntermediaryHolder.Member> entry : holder.getMappings(owner).entrySet()) {
+			IntermediaryHolder.Member member = entry.getValue();
+
+			if (member.isField && fields != null) {
+				// We don't write the debug because it's already been done
+				writeField(sb, owner, member.name, member.descriptor, fields);
+			} else if (methods != null) {
+				writeMethod(sb, owner, member.name, member.descriptor, methods);
+			}
+		}
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/manifest/converter/accesstransformer/AccessTransformerConverter.java]: No need to scan all methods if there are no method wildcards and no need to scan all fields if there are no field wildcards, that would be a good cleanup of the logic (pending the splitting of the mappings in IntermediaryHolder)

File: src/main/java/com/patchworkmc/jar/ForgeModJar.java (hunk 18 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/jar/ForgeModJar.java:30]: Huh, that is true. Then this is only relevant for decompiling then using those transformed mods in development, where dependency AWs are not applied when compiling. But that's not a big deal for now.

File: src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java (hunk 19 above)
Comment (TheGlitch76) [reply src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java:13]: A couple reasons:
- TR's AsmRemapper returns `null` instead of having checked exceptions
-  A TR instance is more expensive to build and requires closing
- We already have to load mappings here for NaiveRemapper, so this saves even more time by reducing the amount of times we load the mappings.

Comment (coderbot16) [reply src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java:13]: My main issue is that we duplicate a bunch of logic probably already contained in tiny-remapper with this new PatchworkRemapper class. I'd much prefer to use the existing code within tiny-remapper instead since it reduces the amount of potential bugs significantly: tiny-remapper is heavily tested. If at all possible I would prefer to use tiny-remapper for these reasons.

Comment (TheGlitch76) [reply src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java:13]: Oh, and this avoids a NaiveRemapper call because ignorefielddesc just... wasn't working or didn't do what the name implies it does

Comment (TheGlitch76) [reply src/main/java/com/patchworkmc/mapping/remapper/PatchworkRemapper.java:13]: Will open an issue when complete

File: src/main/java/com/patchworkmc/mapping/TargetMappingsHolder.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+		Class result = mappings.get(owner);
+		MappingAssertions.assertClassExists(result, owner);
+		return result;
+	}
+
+	/**
+	 * POJO for {@link Member}s.
+	 */
+	public static class Class {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/TargetMappingsHolder.java]: To avoid confusion with the `Class` from Java.
```suggestion
	public static class ClassMembers {
```

File: src/main/java/com/patchworkmc/mapping/TargetMappingsHolder.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+import net.patchworkmc.manifest.accesstransformer.v2.exception.MissingMappingException;
+
+import com.patchworkmc.manifest.converter.accesstransformer.AccessTransformerConverter;
+
+/**
+ * A lazily loading representation of all Minecraft names and descriptors in the target mappings.
+ * Used in {@link AccessTransformerConverter} to resolve descriptors for fields and resolve all names for wildcards.
+ */
+public class TargetMappingsHolder {
```
Comment (coderbot16) [inline src/main/java/com/patchworkmc/mapping/TargetMappingsHolder.java]: I think a name like `MemberInfo` would probably be more indicative of what this is doing, since the purpose of this class is not directly related to mappings / remapping

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Looking mostly good now, just some nitpicks on the class naming.

**Review by coderbot16 — APPROVED** [review]
This looks good! Thanks for undertaking the epic journey of AT support :)
